libsel4muslcsys/CMakeLists.txt

```diff
@@ -73,6 +73,10 @@ set(HighestSyscall 400)
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_XOPEN_SOURCE=700 -DMUSLC_HIGHEST_SYSCALL=${HighestSyscall}")
 
 add_library(sel4muslcsys STATIC EXCLUDE_FROM_ALL ${deps})
+# The init.c is added as an INTERFACE source which forces it to be
+# linked into any dependent binaries. This in turn ensures that the
+# initialisation constructor is included to initialise the C library.
+target_sources(sel4muslcsys INTERFACE "${CMAKE_CURRENT_LIST_DIR}/src/init.c")
```

Comment:
Wow, I didn't know this function existed. (@nomadeel, this could be a better way of doing https://github.com/seL4/util_libs/blob/master/libplatsupport/CMakeLists.txt#L68).
@xurtis, one thing I would be worried about is this file being linked multiple times into a final binary resulting in the same constructor getting called a lot of times.  It's not that uncommon to see `--whole-archive` used as a linker argument to force all symbols from an archive to be linked.  My interpretation of `target_sources` is that this file will get added to every executable and library that depends on sel4muslcsys.  So if you have the following then I would expect this file to get compiled twice:
```
add_libarary(foo SOURCES foo.c)
target_link_libraries(foo sel4muslcsys)

add_executable(bar SOURCES bar.c)
target_link_libraries(bar foo sel4muslcsys)

```

Reply:
If the executable target includes it, the object file will be forced to be included in that target. For each library that includes it, the symbol should only be re-included with that library if it isn't already defined.

Reply:
I'm not sure this would even get included once if it isn't a direct dependency of the executable target

Reply:
The [cmake documentation](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_SOURCES.html#prop_tgt:INTERFACE_SOURCES) gives me the impression that the `INTERFACE` sources will be added to the sources of dependants but not to the `INTERFACE` sources of the dependants, so it shouldn't be transitive in the manner you are suggesting.

Reply:
This also means that if this source is _only_ included transitively, then this translation object won't get included as there are no external references to the function.

Reply:
I tested it and it does indeed behave how I expected:
```
add_library(baz baz2.c)
target_sources(baz INTERFACE ${CMAKE_CURRENT_LIST_DIR}/baz.c)

add_library(foo foo.c)
target_link_libraries(foo baz)

# Depend on baz twice
add_executable(bar bar.c)
target_link_libraries(bar foo baz)

# Use  --whole-archive 
add_executable(bar2 bar.c)
target_link_libraries(bar2 -Wl,--whole-archive foo baz -Wl,--no-whole-archive)

# depend on baz indirectly
add_executable(bar3 bar.c)
target_link_libraries(bar3 foo)

```

```
// baz.c
#define CONSTRUCTOR(priority) __attribute__((constructor(priority)))

#include <stdio.h>
static void CONSTRUCTOR(101) init_hello(void)
{
	printf("hello\n");
} 
```

```
$ ./bar
hello
$ ./bar2
hello
hello
$ ./bar3
hello
```

Reply:
I think placing the contents of init.c into vsyscall.c is a better option here as we already ensure that the constructors in there are unconditionally linked and called.

Reply:
Updated so only an external reference is forced into dependants to ensure that the object with the initialiser is always linked in but the initialiser only ever lives in `sel4muslcsys`.

Reply:
> I think placing the contents of init.c into vsyscall.c is a better option here as we already ensure that the constructors in there are unconditionally linked and called.

We don't though. If none of the functions there are explicitly referenced elsewhere then none of them will be linked. The entire point of this change is to ensure that none of those functions need to be explicitly referenced elsewhere for library initialisation to occur.

Reply:
> We don't though. If none of the functions there are explicitly referenced elsewhere then none of them will be linked. The entire point of this change is to ensure that none of those functions need to be explicitly referenced elsewhere for library initialisation to occur.

We do: https://github.com/seL4/seL4_libs/commit/60f05c39496bd57bf301bab18983f2abc46aac7b